sympy/integrals/integrals.py

```diff
@@ -1242,6 +1242,144 @@ def _sage_(self):
                                     hold=True)
         return f
 
+    def cauchy_principal_value(self, sym, lower_limit=-oo, upper_limit=oo):
+        """
+        Compute the Cauchy Principal Value of the definite integral of a real function in the given interval
+        on the real axis.
+        In mathematics, the Cauchy principal value, is a method for assigning values to certain improper
+        integrals which would otherwise be undefined.
+        The Cauchy principal value of a finite integral of a function f about a point c with a <= c <= b is given by
+                b               c-r                       b
+              /                  /                      /
+             /                  /                      /
+             |                  |                      |
+             |                  |                      |
+        p.v. | f(x)dx  = lim    | f(x)dx   +   lim     |
+             |          r -> 0+ |             r -> 0+  |
+             |                  |                      |
+             /                  /                     /
+            /                  /                     /
+          a                 a                     c+r
+
+        Examples
+        ========
+
+        >>> from sympy import Dummy, symbols, integrate, limit, oo
+        >>> from sympy.integrals.integrals import Integral
+        >>> from sympy.calculus.singularities import singularities
+        >>> x = symbols('x', real = True)
+
+        >>> g = x + 1
+        >>> Integral(g).cauchy_principal_value(x)
+        oo
+
+        >>> f = 1/(x**3)
+        >>> singularities(f,x)
+        {0}
+
+        >>> Integral(f).cauchy_principal_value(x)
+        0
+
+        >>> Integral(f).cauchy_principal_value(x, -oo, 0)
+        -oo
+
+        >>> Integral(f).cauchy_principal_value(x, 0, oo)
+        oo
+
+        >>> Integral(f).cauchy_principal_value(x, 1, 2)
+        3/8
+
+        >>> Integral(f).cauchy_principal_value(x,-2,-1) + Integral(f).cauchy_principal_value(x,1,2)
+        0
+
+        >>> Integral(f).cauchy_principal_value(x, 0.4, 3)
+        3.06944444444444
+
+        >>> Integral(f).cauchy_principal_value(x,-3,-0.4) + Integral(f).cauchy_principal_value(x, 0.4, 3)
+        0
+
+
+        See also
+        ========
+
+        [1] https://en.wikipedia.org/wiki/Cauchy_principal_value
+        [2] http://mathworld.wolfram.com/CauchyPrincipalValue.html
+        """
+        from sympy.calculus import singularities
+
+        self.sym = sym
+        self.lower_limit = lower_limit
+        self.upper_limit = upper_limit
+
+        r = Dummy('r')
+
+        if (self.sym.is_real == False):
```

Comment:
It is not necessary to make assumptions about the integration variable. It will be automatically real on the integration interval.

Reply:
@jksuom  It is necessary. (your previous review :-)
>This is not true in general. Consider, for example, `integrate(1/(1 + x**2), (x, -oo, oo))`.

our main condition is x should be real. In this case, x is complex. That's why there's an error in that case.